include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +262,146 @@ def YieldOp : VerifOp<"yield", [
+      
+    Additional information about how the test should be conducted can be passed 
+    using the two supported attributes. These attributes are `k`, which defines 
+    the bounds of the test, and `type` which defines the type of formal test 
+    being done, i.e. Bounded Model Checking (BMC) or k-Induction.
+  }];
+
+  let arguments = (ins OptionalAttr<APIntAttr>:$k,
+                       OptionalAttr<FormalTestTypeAttr>:$test_type,
+                       OptionalAttr<InnerSymAttr>:$inner_sym);
+
+  let regions = (region SizedRegion<1>:$test_region);
+
+  let assemblyFormat = [{
+    attr-dict  (`sym` $inner_sym^)? `(` (`k` `:` $k^)? (`type` `:` $test_type^)? `)` 
+    $test_region
+  }];
+
+  let extraClassDeclaration = [{
+    /// Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { 
+      return RegionKind::Graph;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Verification Ops
+//===----------------------------------------------------------------------===//
+
+def SymbolicInputOp : VerifOp<"symbolic_input", [HasParent<"verif::FormalOp">]>{
+  let summary = "declare a symbolic value for formal verification";
+  let description = [{
+    This operation declares a symbolic value that can then be used to reason 
+    about `hw.instance` inputs in a symbolic manner in future assertions and 
+    assumptions. The result type must be explicitly marked.  
+  }];
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+def ConcreteValOp : VerifOp<"concrete_input", [
+  AllTypesMatch<["init", "update"]>, HasParent<"verif::FormalOp">
+]> {
+  let summary = "declare a concrete value for formal verification";
+  let description = [{
+    This operation declares a concrete value that can then be used in the 
+    reasoning surrounding symbolic values, allowing for a form of concolic 
+    reasoning to take place in a `verif.formal` block. Concrete values are 
+    defined by an initial value and an update function.
+  }];
+
+  let arguments = (ins AnyType:$init, AnyType:$update);
+  let results = (outs AnyType:$result);
+
+  // Manual definition of a functional type is wanted
+  let assemblyFormat = [{
+    $init `,` $update attr-dict `:` 
+    `(` type($init) `,` type($update) `)` `->` type($result)
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Contract Ops
+//===----------------------------------------------------------------------===//
+
+def ContractOp : VerifOp<"contract", [Pure]>{
+  let summary = "declare formal contract";
+  let description = [{
+    This operation declares a sformal contract that is used to create precondition 
+    and postconditions on a parent module. These are used as an abstraction to 
+    better modularize formal verification such that each module containing a contract 
+    is checked exactly once.
```

Comment:
If a user specifies insufficient postconditions, do we allow whatever verification tool this ends up in to refine its model by exploring the module itself rather than just the post-conditions? Or do we strictly forbid this and just fail to verify?

Reply:
I would say no, the guarantees we give are supposed to be "the module is correct wrt its specification" and not just "the module is correct". So if the user chooses to give a contract, then it is up to them to make it sufficiently tight to be a good abstraction of their system.

Reply:
I agree with @dobios take there.